File: generate_clip_embeddings.py
#!/usr/bin/env python3
"""
Generate CLIP (512-dim) visual embeddings for all product primary images
and store them in product_embeddings.clip_embedding.

Requires the migration SQL to be run first:
  supabase/migrations/20260622213221_add_clip_embedding.sql
"""

import asyncio
import io
import os
import sys
import time
from pathlib import Path

import httpx
import torch
from PIL import Image
from transformers import CLIPModel, CLIPProcessor

# ── Config ──────────────────────────────────────────────────────────────────

SUPABASE_URL = "https://dnghimclwgjmtnesxdmo.supabase.co"
KEY = os.environ['SUPABASE_SERVICE_KEY']

HEADERS = {
    "apikey": KEY,
    "Authorization": f"Bearer {KEY}",
    "Content-Type": "application/json",
}
PATCH_HEADERS = {**HEADERS, "Prefer": "return=minimal"}

CLIP_MODEL = "openai/clip-vit-base-patch32"
CONCURRENCY = 5
IMG_TIMEOUT = 20.0

# ── CLIP setup ───────────────────────────────────────────────────────────────

print(f"Loading {CLIP_MODEL}…")
model = CLIPModel.from_pretrained(CLIP_MODEL)
processor = CLIPProcessor.from_pretrained(CLIP_MODEL)
model.eval()
device = "cpu"  # MPS/CUDA auto-detect if available
if torch.backends.mps.is_available():
    device = "mps"
elif torch.cuda.is_available():
    device = "cuda"
model = model.to(device)
print(f"  CLIP loaded on {device}")


def clip_embed_image(pil_img: Image.Image) -> list[float]:
    inputs = processor(images=pil_img, return_tensors="pt").to(device)
    with torch.no_grad():
        features = model.get_image_features(**inputs)
        features = features / features.norm(dim=-1, keepdim=True)  # L2 normalise
    return features[0].cpu().tolist()


def clip_embed_text(text: str) -> list[float]:
    inputs = processor(text=[text], return_tensors="pt", padding=True).to(device)
    with torch.no_grad():
        features = model.get_text_features(**inputs)
        features = features / features.norm(dim=-1, keepdim=True)
    return features[0].cpu().tolist()


# ── Supabase helpers ─────────────────────────────────────────────────────────

async def fetch_products_needing_clip(client: httpx.AsyncClient) -> list[dict]:
    """
    Return product_embeddings rows where clip_embedding IS NULL,
    joined with primary image URL.
    """
    # Get all embedding rows missing clip_embedding
    r = await client.get(
        f"{SUPABASE_URL}/rest/v1/product_embeddings",
        headers=HEADERS,
        params={
            "select": "id,product_id",
            "clip_embedding": "is.null",
        },
    )
    r.raise_for_status()
    embed_rows = r.json()
    print(f"  {len(embed_rows)} products need CLIP embeddings")
    if not embed_rows:
        return []

    # Fetch primary images for all these product_ids
    product_ids = [str(row["product_id"]) for row in embed_rows]
    chunks = [product_ids[i:i+200] for i in range(0, len(product_ids), 200)]
    image_map = {}  # product_id → url
    for chunk in chunks:
        id_filter = "(" + ",".join(chunk) + ")"
        r2 = await client.get(
            f"{SUPABASE_URL}/rest/v1/product_images",
            headers=HEADERS,
            params={
                "select": "product_id,url",
                "product_id": f"in.{id_filter}",
                "is_primary": "eq.true",
            },
        )
        r2.raise_for_status()
        for img in r2.json():
            image_map[img["product_id"]] = img["url"]

    # Attach image URL to each embedding row; skip if no image
    result = []
    for row in embed_rows:
        url = image_map.get(row["product_id"])
        if url:
            result.append({**row, "image_url": url})
        else:
            print(f"  SKIP product_id={row['product_id']} (no primary image)")
    return result


async def store_clip_embedding(
    client: httpx.AsyncClient, embed_id: int, vector: list[float]
) -> None:
    r = await client.patch(
        f"{SUPABASE_URL}/rest/v1/product_embeddings",
        headers=PATCH_HEADERS,
        params={"id": f"eq.{embed_id}"},
        json={"clip_embedding": vector},
    )
    if r.status_code not in (200, 204):
        raise RuntimeError(f"PATCH {embed_id}: {r.status_code} {r.text[:200]}")


# ── Worker ───────────────────────────────────────────────────────────────────

async def process_row(
    sem: asyncio.Semaphore,
    client: httpx.AsyncClient,
    row: dict,
    idx: int,
    total: int,
) -> None:
    product_id = row["product_id"]
    embed_id   = row["id"]
    img_url    = row["image_url"]

    async with sem:
        try:
            r = await client.get(img_url, timeout=IMG_TIMEOUT, follow_redirects=True)
            r.raise_for_status()
            pil_img = Image.open(io.BytesIO(r.content)).convert("RGB")
            vector  = clip_embed_image(pil_img)
            await store_clip_embedding(client, embed_id, vector)
            print(f"  [{idx}/{total}] ✓ product_id={product_id}")
        except Exception as e:
            print(f"  [{idx}/{total}] ✗ product_id={product_id}: {e}", file=sys.stderr)


# ── Main ─────────────────────────────────────────────────────────────────────

async def main() -> None:
    start = time.monotonic()

    async with httpx.AsyncClient(timeout=60.0) as client:
        rows = await fetch_products_needing_clip(client)
        if not rows:
            print("Nothing to do — all CLIP embeddings already present.")
            return

        sem = asyncio.Semaphore(CONCURRENCY)
        tasks = [
            process_row(sem, client, row, i + 1, len(rows))
            for i, row in enumerate(rows)
        ]
        await asyncio.gather(*tasks)

    elapsed = time.monotonic() - start
    print(f"\nDone in {elapsed:.1f}s")


if __name__ == "__main__":
    asyncio.run(main())

File: generate_clip_embeddings.mjs
/**
 * generate_clip_embeddings.mjs
 * Regenerate CLIP embeddings for all products using the JS Xenova model
 * (CLIPVisionModelWithProjection, fp32) — same vector space as clipEmbed.ts.
 */

import { CLIPVisionModelWithProjection, AutoProcessor, RawImage, env } from "@xenova/transformers";

const SUPABASE_URL = "https://dnghimclwgjmtnesxdmo.supabase.co";
const KEY = process.env.SUPABASE_SERVICE_KEY;
const H     = { apikey: KEY, Authorization: `Bearer ${KEY}`, "Content-Type": "application/json" };
const HMIN  = { ...H, Prefer: "return=minimal" };

const CONCURRENCY = 5;

// ── Load model (once) ────────────────────────────────────────────────────────

console.log("Loading Xenova/clip-vit-base-patch32 (fp32)…");
env.cacheDir = "./.cache/transformers";
const [processor, model] = await Promise.all([
  AutoProcessor.from_pretrained("Xenova/clip-vit-base-patch32"),
  CLIPVisionModelWithProjection.from_pretrained("Xenova/clip-vit-base-patch32", { quantized: false }),
]);
console.log("Model ready.\n");

// ── Supabase helpers ─────────────────────────────────────────────────────────

async function supabaseGetAll(path, params = {}) {
  const PAGE = 1000;
  let offset = 0, all = [];
  while (true) {
    const url = new URL(`${SUPABASE_URL}/rest/v1/${path}`);
    for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
    url.searchParams.set("limit", PAGE);
    url.searchParams.set("offset", offset);
    const r = await fetch(url, { headers: H });
    if (!r.ok) throw new Error(`GET ${path}: ${r.status} ${await r.text()}`);
    const page = await r.json();
    all = all.concat(page);
    if (page.length < PAGE) break;
    offset += PAGE;
  }
  return all;
}

// Keep alias for non-paginated use
async function supabaseGet(path, params = {}) {
  return supabaseGetAll(path, params);
}

async function upsertClipEmbedding(productId, vector) {
  const r = await fetch(`${SUPABASE_URL}/rest/v1/product_embeddings`, {
    method: "PATCH",
    headers: { ...HMIN, Prefer: "return=minimal" },
    body: JSON.stringify({ clip_embedding: vector }),
  });
  // PATCH with query param
  const r2 = await fetch(
    `${SUPABASE_URL}/rest/v1/product_embeddings?product_id=eq.${productId}`,
    { method: "PATCH", headers: HMIN, body: JSON.stringify({ clip_embedding: vector }) }
  );
  if (![200, 204].includes(r2.status)) {
    throw new Error(`PATCH product_id=${productId}: ${r2.status} ${await r2.text()}`);
  }
}

// ── Embed one image URL ───────────────────────────────────────────────────────

async function embedUrl(imageUrl) {
  const image = await RawImage.fromURL(imageUrl);
  const inputs = await processor(image);
  const { image_embeds } = await model(inputs);
  const data = Array.from(image_embeds.data);
  const norm = Math.sqrt(data.reduce((s, v) => s + v * v, 0));
  return data.map(v => v / norm);
}

// ── Build work list ───────────────────────────────────────────────────────────

// 1. All product_embeddings rows (gives us the product_ids to update)
const embedRows = await supabaseGet("product_embeddings", { select: "id,product_id" });
console.log(`product_embeddings rows: ${embedRows.length}`);

// 2. Primary images for all these products (is_primary=true, fallback position=0)
const allProductIds = embedRows.map(r => r.product_id);
const CHUNK = 200;
const imageMap = {};   // product_id → url

for (let i = 0; i < allProductIds.length; i += CHUNK) {
  const chunk = allProductIds.slice(i, i + CHUNK);
  const idFilter = `in.(${chunk.join(",")})`;

  // primary images
  const primaries = await supabaseGet("product_images", {
    select: "product_id,url",
    product_id: idFilter,
    is_primary: "eq.true",
  });
  for (const row of primaries) imageMap[row.product_id] = row.url;

  // fallback: first by position for any still missing
  const missing = chunk.filter(id => !imageMap[id]);
  if (missing.length) {
    const fallbacks = await supabaseGet("product_images", {
      select: "product_id,url",
      product_id: `in.(${missing.join(",")})`,
      order: "position.asc",
      limit: missing.length,
    });
    // keep only first per product
    const seen = new Set();
    for (const row of fallbacks) {
      if (!seen.has(row.product_id)) { imageMap[row.product_id] = row.url; seen.add(row.product_id); }
    }
  }
}

// Build ordered work list, skipping products with no image
const work = embedRows
  .map(r => ({ productId: r.product_id, embedId: r.id, url: imageMap[r.product_id] }))
  .filter(w => w.url);

const skipped = embedRows.length - work.length;
console.log(`Products with images: ${work.length}  (${skipped} skipped — no image)`);
console.log("Starting embedding generation…\n");

// ── Process concurrently ──────────────────────────────────────────────────────

let done = 0, errors = 0;
const start = Date.now();

async function processOne(item) {
  try {
    const vec = await embedUrl(item.url);
    await upsertClipEmbedding(item.productId, vec);
    done++;
    if (done % 50 === 0 || done === work.length) {
      const elapsed = ((Date.now() - start) / 1000).toFixed(0);
      const rate = (done / ((Date.now() - start) / 1000)).toFixed(1);
      console.log(`  [${done}/${work.length}] ${elapsed}s elapsed  ${rate}/s  errors=${errors}`);
    }
  } catch (e) {
    errors++;
    console.error(`  ✗ product_id=${item.productId}: ${e.message}`);
  }
}

// Semaphore via chunked batches
for (let i = 0; i < work.length; i += CONCURRENCY) {
  await Promise.all(work.slice(i, i + CONCURRENCY).map(processOne));
}

const elapsed = ((Date.now() - start) / 1000).toFixed(1);
console.log(`\nDone in ${elapsed}s — ${done} embedded, ${errors} errors, ${skipped} skipped`);
